### ISSUE TITLE
Second SSH shell session to the Linux ssh2 server hangs

### ISSUE THREAD

**Issue by psinghsp** [mscdex/ssh2#239]
For some reason, if we open a second connection to the same Linux server, the library hangs in shell function. Is this a known issue? What can I do to debug this?

   var conn = new Client();

    conn.on('ready', function () {

        self.conn.shell(function (err, stream) { <-- second call hangs on this line
            if (err) {
                throw err;
            }

            self.sshStream = stream;

            stream.setWindow(200, 200, 200, 200);
            stream.on('data', function (data) {
                self.buffer += data.toString('utf8');
            });
            stream.on('close', function(code, signal) {
            });
        });
    });

    conn.connect({
        host: ip,
        port: 22,
        username: user,
        password: password
    })

**Comment by psinghsp** [comment]
The hang is being caused by:
stream.setWindow(200, 200, 200, 200);

If I remove this line, the second ssh session does not hang.

How do I properly set the terminal size for the shell?

**Comment by psinghsp** [comment]
Actually, it hangs randomly. There is no pattern with or without the commented out line. Does anybody else have similar experience with this library?

**Comment by wdavidw** [comment]
Same problem here with version 0.4.4. It hangs randomly. The same code has been running for months with previous 0.3.x versions.

On the target server (centos 6.5 with sshd "OpenSSH_5.3p1, OpenSSL 1.0.1e-fips 11 Feb 2013"), I can read inside "/var/log/secure":
```
Mar 18 14:16:50 master3 sshd[18027]: error: buffer_get_ret: trying to get more bytes 4 than in buffer 0
Mar 18 14:16:50 master3 sshd[18027]: fatal: buffer_get_int: buffer error
```

**Comment by mofux** [comment]
I am getting the same problem. It seems to appear randomly. Looks like the callback for the shell command is never executed.

**Comment by psinghsp** [comment]
Yes - the same experience here. No matter how much I trim the ssh code to bare bones, it just hangs randomly. Do we need to revert and move back to 3.x?

**Comment by mscdex** [comment]
Is this only with `shell()` or does it happen with other session types like `exec()` and such?

**Comment by mofux** [comment]
@psinghsp Can you set ```debug: console.log``` in your ssh2 client options and capture the debug output for a shell attempt that hangs?

@mscdex for me it only seems to happen when I use shell(), I did not see this for exec() - but I don't use exec() excessively.

**Comment by wdavidw** [comment]
I use heavily the exec and sftp functionnality without any problem so yes, only shell

**Comment by wdavidw** [comment]
Considering this code [from this mysql installation procedure line 187](https://github.com/adaltas/node-masson/blob/master/commons/mysql_server.coffee.md) :

```coffee
console.log '*** START ****'
ssh.shell (err, stream) ->
  console.log '*** READY ****'
```

Here's a working debug output:

```
*** START ****
DEBUG: Outgoing: Writing CHANNEL_OPEN (17, session)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: Decrypting
DEBUG: Parser: pktLen:28,padLen:10,remainLen:16
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: Decrypting
DEBUG: Parser: HMAC size:16
DEBUG: Parser: IN_PACKETDATAVERIFY
DEBUG: Parser: Verifying MAC
DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_OPEN_CONFIRMATION
DEBUG: Outgoing: Writing CHANNEL_REQUEST (0, pty-req)
DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: Decrypting
DEBUG: Parser: pktLen:12,padLen:6,remainLen:0
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: HMAC size:16
DEBUG: Parser: IN_PACKETDATAVERIFY
DEBUG: Parser: Verifying MAC
DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_SUCCESS (17)
DEBUG: Outgoing: Writing CHANNEL_REQUEST (0, shell)
DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: Decrypting
DEBUG: Parser: pktLen:28,padLen:18,remainLen:16
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: Decrypting
DEBUG: Parser: HMAC size:16
DEBUG: Parser: IN_PACKETDATAVERIFY
DEBUG: Parser: Verifying MAC
DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_WINDOW_ADJUST (17, 2097152)
DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: Decrypting
DEBUG: Parser: pktLen:12,padLen:6,remainLen:0
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: HMAC size:16
DEBUG: Parser: IN_PACKETDATAVERIFY
DEBUG: Parser: Verifying MAC
DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_SUCCESS (17)
*** READY ****
```

And here's a failed execution (same code, retried 5-10 times before getting it):

```
*** START ****
DEBUG: Outgoing: Writing CHANNEL_OPEN (17, session)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: Decrypting
DEBUG: Parser: pktLen:28,padLen:10,remainLen:16
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: Decrypting
DEBUG: Parser: HMAC size:16
DEBUG: Parser: IN_PACKETDATAVERIFY
DEBUG: Parser: Verifying MAC
DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_OPEN_CONFIRMATION
DEBUG: Outgoing: Writing CHANNEL_REQUEST (0, pty-req)
DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
DEBUG: Outgoing: Writing CHANNEL_CLOSE (0)
```

Note, the exact same messages until the last line with "CHANNEL_CLOSE".

**Comment by thingalon** [comment]
I was able to replicate the issue, tracked it down to a problem in ssh-streams. 

Pull request here: https://github.com/mscdex/ssh2-streams/pull/5

**Comment by mscdex** [comment]
Ok I've published `ssh2-streams` 0.0.8. Can others that were encountering this problem please test by re-installing `ssh2` to pick up the new `ssh2-streams`?

**Comment by wdavidw** [comment]
seems ok now, thanks
